### DISCUSSION TITLE
definitelytyped including all dependencies

### DISCUSSION THREAD

**Discussion by jfairley** [renovatebot/renovate#9807] [Request Help]
cloned from https://github.com/renovatebot/app-support/issues/119

<hr>

Recently, I've noticed that my repos have a renovate PR seeming to be grouped as `definitelytyped`, but the PR seems to include _every_ dependency update that my app needs.

I'm not sure whether this is an issue with the shareable configs or one of my configs, but I'll share some of my configs so we can figure it out...

This worked fine for a long time until recently, and we don't change our config often, which is why I'm suspicious of a shareable config.

<hr>

First, the pull request... 😱 

The title here is `fix(deps): update definitelytyped`.

![renovate-definitelytyped](https://user-images.githubusercontent.com/992483/116735906-67732300-a9ac-11eb-815e-86fcc3b8686e.gif)

(Sorry for the low res. Github limits uploaded files to 10MB. It still gets the point across.)

<hr>

Next, an excerpt from my config:

```json
{
  "extends": [
    "config:base",
    "docker:pinDigests",
    "group:recommended",
    "group:reactMonorepo",
    "schedule:weekly"
  ],


  "packageRules": [


    {
      "automerge": true,
      "excludePackageNames": ["@types/react-table"],
      "extends": ["group:definitelyTyped"]
    },


  ]
}
```

<details>
<summary>Full config</summary>

```json
{
  "extends": [
    "config:base",
    "docker:pinDigests",
    "group:recommended",
    "group:reactMonorepo",
    "schedule:weekly"
  ],
  "encrypted": {
    "npmToken": "-- redacted --"
  },
  "npmrc": "@adcellerant:registry=https://npm.pkg.github.com/\n//npm.pkg.github.com/:_authToken=${NPM_TOKEN}\nalways-auth=true",
  "commitBody": "[skip release]",
  "labels": ["dependencies"],
  "packageRules": [
    {
      "automerge": true,
      "datasources": ["docker", "orb"],
      "groupName": "containers"
    },
    {
      "automerge": true,
      "depTypeList": ["devDependencies"],
      "updateTypes": ["minor", "patch", "pin", "digest"]
    },
    {
      "groupName": "dev-tools",
      "packageNames": [
        "check-node-version",
        "concurrently",
        "create-ts-index",
        "fixpack",
        "husky",
        "is-ci",
        "lint-staged",
        "nodemon",
        "prettier",
        "rimraf",
        "semantic-release",
        "ts-node",
        "wait-on"
      ],
      "packagePatterns": ["^@commitlint/", "^@semantic-release/", "webpack"]
    },
    {
      "automerge": true,
      "groupName": "cypress",
      "packageNames": ["cypress"],
      "packagePatterns": ["^@cypress/", "^cypress-"]
    },
    {
      "automerge": true,
      "excludePackageNames": ["@types/react-table"],
      "extends": ["group:definitelyTyped"]
    },
    {
      "automerge": true,
      "groupName": "enzyme",
      "packageNames": ["enzyme"],
      "packagePatterns": ["^enzyme-"]
    },
    {
      "automerge": true,
      "extends": ["group:fortawesome"]
    },
    {
      "automerge": true,
      "groupName": "graphql",
      "packagePatterns": ["apollo", "graphql"]
    },
    {
      "automerge": true,
      "groupName": "jest",
      "packageNames": ["@types/jest", "jest", "ts-jest"],
      "packagePatterns": ["^jest-"]
    },
    {
      "automerge": true,
      "extends": ["group:linters"]
    },
    {
      "automerge": true,
      "groupName": "styled-components",
      "packagePatterns": ["styled-components"]
    },
    {
      "automerge": true,
      "groupName": "supertest",
      "packageNames": ["supertest"]
    },
    {
      "automerge": true,
      "excludePackageNames": ["react-table"],
      "packageNames": [
        "@welldone-software/why-did-you-render",
        "fast-memoize",
        "formik",
        "lodash",
        "micro-memoize",
        "moment",
        "tslib",
        "typescript",
        "uuid",
        "validator",
        "victory",
        "w3c-keys",
        "winston"
      ],
      "packagePatterns": ["react"]
    }
  ],
  "separateMultipleMajor": true,
  "timezone": "America/Denver",
  "updateNotScheduled": false
}
```

</details>

<hr>

🤔

**Comment by rarkins** [comment]
As mentioned in the other repo, this is actually an invalid config, however I've worked out a better way to migrate it: #9814

**Reply by jfairley** [reply]
@rarkins my config is set to run weekly, and based on last night's run, I do think this is a bug or bad shared config, probably coming from `"group:recommended"`, because I can't reason where else would be the source.

I did make the changes you recommended, btw.

<details>
<summary>Here's my latest config, which I updated on Friday.</summary>

```json
{
  "extends": [
    "config:base",
    "docker:pinDigests",
    "group:recommended",
    "group:reactMonorepo",
    "schedule:weekly"
  ],
  "encrypted": {
    "npmToken": "<redacted>"
  },
  "npmrc": "@adcellerant:registry=https://npm.pkg.github.com/\n//npm.pkg.github.com/:_authToken=${NPM_TOKEN}\nalways-auth=true",
  "commitBody": "[skip release]",
  "labels": ["dependencies"],
  "packageRules": [
    {
      "automerge": true,
      "datasources": ["docker", "orb"],
      "groupName": "containers"
    },
    {
      "automerge": true,
      "depTypeList": ["devDependencies"],
      "updateTypes": ["minor", "patch", "pin", "digest"]
    },
    {
      "groupName": "dev-tools",
      "matchPackageNames": [
        "check-node-version",
        "concurrently",
        "create-ts-index",
        "fixpack",
        "husky",
        "is-ci",
        "lint-staged",
        "nodemon",
        "rimraf",
        "ts-node",
        "wait-on"
      ],
      "matchPackagePatterns": [
        "commitlint",
        "prettier",
        "semantic-release",
        "webpack"
      ]
    },
    {
      "automerge": true,
      "groupName": "cypress",
      "matchPackageNames": ["cypress"],
      "matchPackagePrefixes": ["@cypress/", "cypress-"]
    },
    {
      "automerge": true,
      "groupName": "definitelyTyped",
      "matchPackagePrefixes": ["@types/"],
      "excludePackageNames": ["@types/react-table"]
    },
    {
      "automerge": true,
      "groupName": "enzyme",
      "matchPackageNames": ["enzyme"],
      "matchPackagePrefixes": ["enzyme-"]
    },
    {
      "automerge": true,
      "extends": ["group:fortawesome"]
    },
    {
      "automerge": true,
      "groupName": "graphql",
      "matchPackagePatterns": ["apollo", "graphql"]
    },
    {
      "automerge": true,
      "groupName": "jest",
      "matchPackageNames": ["@types/jest", "jest", "ts-jest"],
      "matchPackagePrefixes": ["jest-"]
    },
    {
      "automerge": true,
      "extends": ["group:linters"]
    },
    {
      "automerge": true,
      "groupName": "styled-components",
      "matchPackagePatterns": ["styled-components"]
    },
    {
      "automerge": true,
      "groupName": "supertest",
      "matchPackageNames": ["supertest"]
    },
    {
      "automerge": true,
      "matchPackageNames": [
        "@simbathesailor/use-what-changed",
        "@welldone-software/why-did-you-render",
        "fast-memoize",
        "formik",
        "lodash",
        "micro-memoize",
        "moment",
        "tslib",
        "typescript",
        "uuid",
        "validator",
        "victory",
        "w3c-keys",
        "winston"
      ],
      "matchPackagePatterns": ["react"],
      "excludePackageNames": ["react-table"]
    }
  ],
  "separateMultipleMajor": true,
  "timezone": "America/Denver",
  "updateNotScheduled": false
}
```
</details>

Renovate ran in the AM this morning. I current have a total of 35 "definitelytyped" PRs being maintained across my org repos. Focusing on just 1 of those repos, here are 3 active "definitelytyped" PRs (in the same repo) split up by the type of major they're trying to upgrade.


(all outdated dependencies that are not a major upgrade) 

<img width="941" alt="Screen Shot 2021-05-03 at 9 06 57 AM" src="https://user-images.githubusercontent.com/992483/116894172-012a1280-abef-11eb-8d5f-31b7ce254b1a.png">

(all outdated dependencies that are a major upgrade v10) 

<img width="943" alt="Screen Shot 2021-05-03 at 9 07 05 AM" src="https://user-images.githubusercontent.com/992483/116894175-025b3f80-abef-11eb-8f51-a9b40614ebf5.png">

(all outdated dependencies that are a major upgrade v4) 

<img width="942" alt="Screen Shot 2021-05-03 at 9 07 14 AM" src="https://user-images.githubusercontent.com/992483/116894178-02f3d600-abef-11eb-8e54-adb054116917.png">

**Reply by jfairley** [reply]
Notice that I have a config titled `definitelyTyped` (uppercase T), but the PRs are grouped as `definitelytyped` (lowercase T).

**Reply by rarkins** [reply]
Can you reproduce this publicly?

**Reply by jfairley** [reply]
Not easily. I'd have to copy these configs to a public repo, since all of my org repos are private.

**Reply by jfairley** [reply]
I'll give it a shot though. It'll be at least a few hours to get through some other work first.

**Reply by jfairley** [reply]
@rarkins I've created the repo https://github.com/jfairley/renovate-test-repo. Initially, with some basic updates, I wasn't able to reproduce the issue. 

With my last commit, I've copied my entire private config minus the github packages auth. The following issue was opened: 

> There is an error with this repository's Renovate configuration that needs to be fixed. As a precaution, Renovate will stop PRs until it is resolved.
> 
> Location: renovate.json
Error type: The renovate configuration file contains some invalid settings
Message: packageRules[6]: Each packageRule must contain at least one match* or exclude* selector. Rule: {"automerge":true,"extends":["group:fortawesome"]}, packageRules[9]: Each packageRule must contain at least one match* or exclude* selector. Rule: {"automerge":true,"extends":["group:linters"]}

1. Why is this issue was not opened in my private repos?
2. Does a package rule missing "match*" and "exclude*" fallback to "match all"?
3. _If this is the issue_, how would it manifest as a bunch of PRs titled `definitelytyped` since I _do_ (as of Friday) have a "match*" _and_ "exclude*" statement on my only "definitelyTyped" package rule.

**Reply by rarkins** [reply]
As a result of this discussion, this commit was put live in the app about 12 hours ago: https://github.com/renovatebot/renovate/commit/a25eca7cb139a9c754e990447e62a3777fa03651

So that possibly explains (1). Another explanation for (1) can be that if the config problem is nested within a preset then it might not be picked up the same way (we do try to though).

(2) No, it should not fall back to that. A package rule which never matches should end up having no effect at all.

(3) I can't say for sure yet. But if you're extending `group:definitelyTyped` within a package rule then it's definitely related.

**Reply by jfairley** [reply]
1. Looking at that test case for nested package rules, when I have something like `"extends": "group:definitelyTyped"` in a package rule, is that what I end up with after the config is parsed (a nested package rule)?

2. Does this process run every time renovate parses my config? In other words, do I need to do anything to my config to pick up the change you referenced, or will it apply the next time renovate goes to run?

I'm going to request a rebase on one of those PRs and see if anything changes.

**Reply by rarkins** [reply]
1. Yes,  that's what causes the nesting.
2. Yes, every time

The best thing to do would be to rewrite the affected packageRules so that you don't have any `"extends": ["group:x"]` inside packageRules. I can help you here with any you're not sure of. Sometimes the groups have a corresponding `monorepo:x` which you *can* extend.

**Reply by jfairley** [reply]
One that would be somewhat painful to not use is

```json
    {
      "automerge": true,
      "extends": ["group:linters"]
    },
```

`group:linters` that you have is really valuable to me, and all I want to add (clearly) is allowing it to automerge.

Trying to think of different ways to pull this off, what if I pulled these automerge-able groups out to a separate config like so:

```json
# automerge-groups.json
{
  "extends": [
    "group:fontawesome",
    "group:linters",
  ],
  "automerge": true
}
```

And secondly, what would happen if my `default.json` extends that? More pointedly, would having `"automerge": true` at the root of `automerge-groups.json` implicitly enable automerge for `default.json`?

```json
# default.json
{
  "extends": [
    "github>AdCellerant/renovate-config:automerge-groups",
    "group:recommended",   # <--- things I don't intend to automerge
  ],
}
```
